src/net.cpp

```diff
@@ -678,14 +679,40 @@ int CNetMessage::readData(const char *pch, unsigned int nBytes)
     return nCopy;
 }
 
-const uint256& CNetMessage::GetMessageHash() const
+const uint256& TransportDeserializer::GetMessageHash() const
 {
     assert(complete());
     if (data_hash.IsNull())
         hasher.Finalize(data_hash.begin());
     return data_hash;
 }
 
+CNetMessage TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) {
+    // decompose a single CNetMessage from the TransportDeserializer
+    CNetMessage msg(vRecv);
+
+    // store state about valid header, netmagic and checksum
+    msg.m_valid_header = hdr.IsValid(message_start);
+    msg.m_valid_netmagic = (memcmp(hdr.pchMessageStart, message_start, CMessageHeader::MESSAGE_START_SIZE) == 0);
```

Comment:
Isn't this check redundant with first check in `IsValid` ? Furthermore isn't netmagic validity a subset of header one,  and so you may remove it completely and only rely on `m_valid_header` ?

Reply:
Yes. It's redundant for the same reason as said above. Further optimizations would change the behavior of v1 decomposing.

`ProcessMessages()` happens after de-queuing deserialized messages. If we would check during deserialization (and disconnect), we would change the behavior.

Also, `ProcessMessages()` differentiates between invalid netmagic and invalid header. It tolerates invalid headers (while not invalid netmagics), thus we need a way to transport those check-states without breaking the current behavior.

Since this PR aims to do a refactoring, I don't think behavioral changes are ideal here.

Reply:
Ah yes exact, you may have a valid netmagic and an invalid header and get accepted. 

Okay, I agree if these points are addressed later in behavioral changes specific PRs that's better than everything at once.